### ISSUE TITLE
[Goa V3 - Fixed in 3.1.2] SkipRequestBodyEncodeDecode and JWTAuth generates broken code

### ISSUE THREAD

**Issue by saniales** [goadesign/goa#2536]
***EDIT: FIXED IN 3.1.2***, see last message

Here an example, if you happen to create the following Method

``` go
var JWTAuth = JWTSecurity("jwt", func() {
	Description(`Secures endpoint by requiring a valid JWT token retrieved via the signin endpoint. Supports scopes "api:read" and "api:write".`)
	Scope("api:read", "Read-only access")
	Scope("api:write", "Read and write access")
})

// ...
Service("test", func() {
    Method("upload-profile-image", func() {
        Security(JWTAuth) // IMPORTANT LINE

	Payload(func() {
		TokenField(1, "token", String, func() { // IMPORTANT LINE
			Description("JWT used for authentication")
		})
		Field(2, "content_type", String, "Content-Type header, must define value for multipart boundary.", func() {
			Default("multipart/form-data; boundary=-BLAHBLAH---")
			Pattern("multipart/[^;]+; boundary=.+")
			Example("multipart/form-data; boundary=-BLAHBLAH---")
		})
	})

        Error("unauthorized", ErrorResult, "JWT token is invalid") // IMPORTANT LINE
	Error("invalid_media_type", ErrorResult, "Error returned when the Content-Type header does not define a multipart request.")
	Error("invalid_multipart_request", ErrorResult, "Error returned when the request body is not a valid multipart content.")
	Error("internal_error", ErrorResult, "Fault while processing upload.")

	HTTP(func() {
		Header("token:Authorization") // IMPORTANT LINE
		Header("content_type:Content-Type")

		POST("/")

		SkipRequestBodyEncodeDecode() // IMPORTANT LINE

		// Define error HTTP statuses.
		Response("unauthorized", StatusUnauthorized)
		Response("invalid_media_type", StatusBadRequest)
		Response("invalid_multipart_request", StatusBadRequest)
		Response("internal_error", StatusInternalServerError)
	})
    })
})
```

Running `goa gen my/awesome/packagename/design` and later

Running `goa example my/awesome/packagename/design`

and building gives the following error

```
# my/awesome/packagename/gen/test
gen/test/endpoints.go:76:6: undefined: p
gen/test/endpoints.go:77:13: undefined: p
```

which corresponds to the generated code

``` go 
// gen/test/endpoints.go section
// NewUploadProfileImageEndpoint returns an endpoint function that calls the
// method "upload-profile-image" of service "dr-coder".
func NewUploadProfileImageEndpoint(s Service, authJWTFn security.AuthJWTFunc) goa.Endpoint {
	return func(ctx context.Context, req interface{}) (interface{}, error) {
		ep := req.(*UploadProfileImageRequestData)
		var err error
		sc := security.JWTScheme{
			Name:           "jwt",
			Scopes:         []string{"api:read", "api:write"},
			RequiredScopes: []string{},
		}
		var token string
		if p.Token != nil { // ERROR
			token = *p.Token // ERROR
		}
		ctx, err = authJWTFn(ctx, token, &sc)
		if err != nil {
			return nil, err
		}
		return nil, s.UploadProfileImage(ctx, ep.Payload, ep.Body)
	}
}
```

this can be solved by changing the generated code in the following way when there is `SkipRequestBodyEncodeDecode` (therefore I have a change in param due to the fact that i am using the req directly)

``` go
// NewUploadProfileImageEndpoint returns an endpoint function that calls the
// method "upload-profile-image" of service "test".
func NewUploadProfileImageEndpoint(s Service, authJWTFn security.AuthJWTFunc) goa.Endpoint {
	return func(ctx context.Context, req interface{}) (interface{}, error) {
		ep := req.(*UploadProfileImageRequestData)
		var err error
		sc := security.JWTScheme{
			Name:           "jwt",
			Scopes:         []string{"api:read", "api:write"},
			RequiredScopes: []string{},
		}
		var token string
		if ep.Payload.Token != nil { // p -> ep.Payload
			token = *ep.Payload.Token // p-> ep.Payload
		}
		ctx, err = authJWTFn(ctx, token, &sc)
		if err != nil {
			return nil, err
		}
		return nil, s.UploadProfileImage(ctx, ep.Payload, ep.Body)
	}
}
```

I hope it helps, for now I have to edit the generated code, which isn't nice. Please fix

**Comment by saniales** [comment]
This seems to be connected to the following files in [security_functions.go](https://github.com/goadesign/goa/blob/8ecad6641506a90dcb4ccdc15c56e050f1cf6f3c/codegen/service/testdata/security_functions.go)

https://github.com/goadesign/goa/blob/8ecad6641506a90dcb4ccdc15c56e050f1cf6f3c/codegen/service/testdata/security_functions.go#L57-L60

is there any way to parametrize this with templates like the endpoint function? @raphael 

``` go
var code = `var token string 
if {{if .SkipRequestBodyEncodeDecode}}ep{{else}}p{{end}}.Token != nil { 
    token = *{{if .SkipRequestBodyEncodeDecode}}ep{{else}}p{end}}.Token 
}`
```

**Comment by saniales** [comment]
Upon better search the incriminated code is

https://github.com/goadesign/goa/blob/8ecad6641506a90dcb4ccdc15c56e050f1cf6f3c/codegen/service/endpoint.go#L228-L235

in particular, the  `{{- $payload := payloadVar . }}`

I don't understand, expecially because of 
https://github.com/goadesign/goa/blob/8ecad6641506a90dcb4ccdc15c56e050f1cf6f3c/codegen/service/endpoint.go#L158-L160

that should populate payloadVar properly

**Comment by saniales** [comment]
IN branch V2 the update of service/endpoint.go has not been reported, and this may be the cause of the issue

**Comment by saniales** [comment]
from 

https://github.com/goadesign/goa/commit/8ecad6641506a90dcb4ccdc15c56e050f1cf6f3c

it seems that this issue is fixed in 3.1.2

solved by using that version

in `go.mod`

```
	goa.design/goa/v3 v3.1.2-0.20200419184530-8ecad6641506
```

**Comment by raphael** [comment]
Indeed, this was just fixed. Will be releasing v3.1.2 with the fix by next week.
